### DISCUSSION TITLE
Training GlowTTS model using Common Voice dataset from Mozilla for bengali TTS

### DISCUSSION THREAD

**Discussion by ianujkr** [coqui-ai/TTS#3679] [General Q&A]
I am trying to train a glow TTS model for bengali langauge. The dataset contains 21000 wav file and a train.tsv file. However during the initial phase of training I am facing an error:

| > Max audio length: 339862.0
 | > Min audio length: 35734.0
 | > Avg audio length: 183501.73505900268
 | > Num. instances discarded samples: 0
 | > Batch group size: 0.
**/usr/lib/python3.10/multiprocessing/popen_fork.py:66: RuntimeWarning: os.fork() was called. os.fork() is incompatible with multithreaded code, and JAX is multithreaded, so this will likely lead to a deadlock.**
  self.pid = os.fork()
 ! Run is removed from /content/dataset/cv-corpus-17.0-2024-03-15/run-April-11-2024_07+50AM-0000000
Internal Python error in the inspect module.
Below is the traceback from this internal error.

Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/trainer/trainer.py", line 1833, in fit
    self._fit()
  File "/usr/local/lib/python3.10/dist-packages/trainer/trainer.py", line 1785, in _fit
    self.train_epoch()
  File "/usr/local/lib/python3.10/dist-packages/trainer/trainer.py", line 1503, in train_epoch
    for cur_step, batch in enumerate(self.train_loader):
  File "/usr/local/lib/python3.10/dist-packages/torch/utils/data/dataloader.py", line 631, in __next__
    data = self._next_data()
  File "/usr/local/lib/python3.10/dist-packages/torch/utils/data/dataloader.py", line 1346, in _next_data
    return self._process_data(data)
  File "/usr/local/lib/python3.10/dist-packages/torch/utils/data/dataloader.py", line 1372, in _process_data
    data.reraise()
  File "/usr/local/lib/python3.10/dist-packages/torch/_utils.py", line 722, in reraise
    raise exception
AssertionError: Caught AssertionError in DataLoader worker process 0.
Original Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/torch/utils/data/_utils/worker.py", line 308, in _worker_loop
    data = fetcher.fetch(index)
  File "/usr/local/lib/python3.10/dist-packages/torch/utils/data/_utils/fetch.py", line 51, in fetch
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/usr/local/lib/python3.10/dist-packages/torch/utils/data/_utils/fetch.py", line 51, in <listcomp>
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/usr/local/lib/python3.10/dist-packages/TTS/tts/datasets/dataset.py", line 201, in __getitem__
    return self.load_data(idx)
  File "/usr/local/lib/python3.10/dist-packages/TTS/tts/datasets/dataset.py", line 257, in load_data
    token_ids = self.get_token_ids(idx, item["text"])
  File "/usr/local/lib/python3.10/dist-packages/TTS/tts/datasets/dataset.py", line 240, in get_token_ids
    token_ids = self.get_phonemes(idx, text)["token_ids"]
  File "/usr/local/lib/python3.10/dist-packages/TTS/tts/datasets/dataset.py", line 219, in get_phonemes
    assert len(out_dict["token_ids"]) > 0 

#The code for the training is 

`from TTS.tts.datasets import load_tts_samples
from TTS.tts.models.glow_tts import GlowTTS
from TTS.tts.utils.text.tokenizer import TTSTokenizer
from TTS.utils.audio import AudioProcessor

import os

subset_dataset_config = BaseDatasetConfig(
    formatter="common_voice",
    meta_file_train="train.tsv",
    path=os.path.join("/content/dataset/cv-corpus-17.0-2024-03-15/bn")
)
subset_dataset_config.max_samples = 1000

output_path = "/content/dataset/cv-corpus-17.0-2024-03-15"

# GlowTTSConfig: all model related values for training, validating and testing.
from TTS.tts.configs.glow_tts_config import GlowTTSConfig
config = GlowTTSConfig(
    batch_size=32,
    eval_batch_size=16,
    num_loader_workers=4,
    num_eval_loader_workers=4,
    run_eval=True,
    test_delay_epochs=-1,
    epochs=2,
    text_cleaner="phoneme_cleaners",
    use_phonemes=True,
    phoneme_language="bn",
    phoneme_cache_path=os.path.join(output_path, "phoneme_cache"),
    print_step=25,
    print_eval=False,
    mixed_precision=True,
    output_path=output_path,
    datasets=[subset_dataset_config],
    save_step=1000,
)

from TTS.utils.audio import AudioProcessor
ap = AudioProcessor.init_from_config(config)

from TTS.tts.utils.text.tokenizer import TTSTokenizer
tokenizer, config = TTSTokenizer.init_from_config(config)

from TTS.tts.datasets import load_tts_samples
train_samples, eval_samples = load_tts_samples(
    subset_dataset_config,
    eval_split=True,
    eval_split_max_size=config.eval_split_max_size,
    eval_split_size=config.eval_split_size,
)

from TTS.tts.models.glow_tts import GlowTTS
model = GlowTTS(config, ap, tokenizer, speaker_manager=None)

!pip3 install tensorflow==2.9.1

from trainer import Trainer, TrainerArgs
trainer = Trainer(
    TrainerArgs(),
    config,
    output_path,
    model=model,
    train_samples=train_samples,
    eval_samples=eval_samples
)

trainer.fit()`

I am using the Colab Free based on T4 GPU for training. However my training cannot complete due to above reason i.e. "/usr/lib/python3.10/multiprocessing/popen_fork.py:66: RuntimeWarning: os.fork() was called. os.fork() is incompatible with multithreaded code, and JAX is multithreaded, so this will likely lead to a deadlock." Please help me resolve the above issue.

**Comment by eginhard** [comment]
The error mentions `os.fork()` and Jax, both of which aren't used in Coqui TTS, so not sure why these are in your environment? And why are you also installing Tensorflow?

**Reply by ianujkr** [reply]
If I don't install the tensorflow, it is giving me the error as follows:

**Error**
ImportError                               Traceback (most recent call last)
[/usr/local/lib/python3.10/dist-packages/tensorboard/compat/__init__.py](https://localhost:8080/#) in tf()
     41     try:
---> 42         from tensorboard.compat import notf  # noqa: F401
     43     except ImportError:

ImportError: cannot import name 'notf' from 'tensorboard.compat' (/usr/local/lib/python3.10/dist-packages/tensorboard/compat/__init__.py)

During handling of the above exception, another exception occurred:

RuntimeError                              Traceback (most recent call last)
RuntimeError: module compiled against API version 0x10 but this version of numpy is 0xf
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
[/usr/local/lib/python3.10/dist-packages/tensorboard/compat/__init__.py](https://localhost:8080/#) in tf()
     41     try:
---> 42         from tensorboard.compat import notf  # noqa: F401
     43     except ImportError:

ImportError: cannot import name 'notf' from 'tensorboard.compat' (/usr/local/lib/python3.10/dist-packages/tensorboard/compat/__init__.py)

During handling of the above exception, another exception occurred:

SystemError                               Traceback (most recent call last)
25 frames
[/usr/local/lib/python3.10/dist-packages/tensorflow/python/training/py_checkpoint_reader.py](https://localhost:8080/#) in <module>
     17 from tensorflow.python.framework import errors_impl
     18 from tensorflow.python.util import compat
---> 19 from tensorflow.python.util._pywrap_checkpoint_reader import CheckpointReader
     20 from tensorflow.python.util.tf_export import tf_export
     21 

SystemError: initialization of _pywrap_checkpoint_reader raised unreported exception

To deal with the following error the GPT suggested me to install tensorboard.
